core/distributed/matrix.cpp

```diff
@@ -302,8 +312,16 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
         ->read(std::move(non_local_data));
 
     // exchange step 1: determine recv_sizes, send_sizes, send_offsets
-    exec->get_master()->copy_from(
-        exec, num_parts, recv_sizes_array.get_const_data(), recv_sizes_.data());
+    auto host_recv_targets =
+        make_temporary_clone(exec->get_master(), &imap.get_remote_target_ids());
+    auto host_offsets = make_temporary_clone(
+        exec->get_master(), &imap.get_remote_global_idxs().get_offsets());
+    std::fill(recv_sizes_.begin(), recv_sizes_.end(), 0);
+    for (size_type i = 0; i < host_recv_targets->get_size(); ++i) {
+        recv_sizes_[host_recv_targets->get_const_data()[i]] =
+            host_offsets->get_const_data()[i + 1] -
+            host_offsets->get_const_data()[i];
+    }
```

Comment:
It feels pedantic, but I think having a raw loop next to a bunch of algorithm calls is a mismatch of abstraction levels - this might be better extracted into a separate function, similar to Sean Parent's "no raw loops" suggestion (https://www.youtube.com/watch?v=W2tWOdzgXHA)

Maybe we can call it `compute_recv_sizes`? Then the high-level control flow is easy to follow, and if you need to know how those recv_sizes get determined, you look at the function call.

Reply:
IMO the issue here is that arrays can't be used with STL. If we had begin, and end for them, this would just be `std::adjacent_difference`

Reply:
You can use pointers here , but `adjacent_difference` is not 100% sufficient, since it leaves the first element unchanged and doesn't scatter the results. I just meant extract these 5 lines into a function

Reply:
You're right, `adjacent_difference` is weird, same as `partial_sum`. I would still argue that the issue here is just that our arrays are not easy to access. If it would read
```
    for (size_type i = 0; i < ecv_targets.size(); ++i) {
        recv_sizes_[recv_targets[i]] = offsets[i + 1] - offsets[i];
    }
```
I don't think you would have the same issue. Only because we have to clutter it with `host_` prefixes, and use very verbose function names it becomes an issue.

Reply:
My comment is mostly not about the complexity of the code (`.get_const_data()[]` vs `[]`) but about mixing low-level code with high-level code. I do agree that it would be nice to have a cleaner access to arrays. If your extracted function takes pointers, you would arrive at that code though.

Maybe we could add a `host_array` type that is derived from an `array`, just with `operator[]` and `begin/end()` functions?

Reply:
I've changed this, but maybe not exactly how you imagined it. I'm using a lambda for this, since I don't want to put it outside the function, due to it's limited use.